lib/src/sink_base.dart

```diff
@@ -0,0 +1,168 @@
+// Copyright (c) 2021, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import 'dart:async';
+import 'dart:convert';
+
+import 'package:meta/meta.dart';
+
+import 'async_memoizer.dart';
+
+/// An abstract class that implements [EventSink] in terms of [onData],
+/// [onError], and [onClose] methods.
+///
+/// This takes care of ensuring that events can't be added after [close] is
+/// called.
+abstract class EventSinkBase<T> implements EventSink<T> {
+  /// Whether [close] has been called and no more events may be written.
+  bool get _closed => _closeMemo.hasRun;
+
+  @override
+  void add(T data) {
+    _checkCanAddEvent();
+    onAdd(data);
+  }
+
+  /// A method that handles data events that are passed to the sink.
+  @visibleForOverriding
+  void onAdd(T data);
+
+  @override
+  void addError(Object error, [StackTrace? stackTrace]) {
+    _checkCanAddEvent();
+    onError(error, stackTrace);
+  }
+
+  /// A method that handles error events that are passed to the sink.
+  @visibleForOverriding
+  void onError(Object error, [StackTrace? stackTrace]);
+
+  @override
+  Future<void> close() => _closeMemo.runOnce(onClose);
+  final _closeMemo = AsyncMemoizer<void>();
+
+  /// A method that handles the sink being closed.
+  ///
+  /// This may return a future that completes once the stream sink has shut
+  /// down. If cleaning up can fail, the error may be reported in the returned
+  /// future.
+  @visibleForOverriding
+  FutureOr<void> onClose();
+
+  /// Asserts that the sink is in a state where adding an event is valid.
+  void _checkCanAddEvent() {
+    if (_closed) throw StateError('Cannot add event after closing');
+  }
+}
```

Comment:
I think it would be better to have a `SinkBase` which takes `onData`/`onError`/`onDone` as function arguments, or as a single object (could perhaps even be a `StreamSubscription`-like object).

Don't mingle the sink and the response APIs into one interface.

Don't use `@visibleForOverriding`. We generally don't use annotations from `package:meta` in the platform-adjacent packages. (Also not necessary if you don't merge  the interfaces)

Reply:
Why not use `package:meta` and `@visibleForOverriding` if they result in better APIs? Using inheritance requires the creation of fewer intermediate objects, which results in less complexity and more efficiency.

"Platform-adjacent packages" are only what you declare them to be, and `meta` is certainly a reasonable inclusion since it tightly integrates into the platform.

Reply:
I think annotation-based "language extensions" like these are fine for applications, but also that they do not belong in reusable package APIs. Such an API should not expose something that you are not intending to *actually* expose, because users can, and will, use them as if they were just public.
So, it's not a *better* API to me, it's a messier and less structured API, one which mixes different responsibilities in the same interface.